packages/ui/src/components/Document/actions/AttachSchemaButton.tsx

```diff
@@ -62,25 +114,92 @@ export const AttachSchemaButton: FunctionComponent<AttachSchemaProps> = ({
     } finally {
       setIsLoading(false);
     }
+    setIsModalOpen(false);
   }, [
     addAlert,
     api,
     clearNodeReferencesForDocument,
     documentId,
     documentType,
+    filePaths,
     reloadNodeReferences,
     setIsLoading,
     updateDocumentDefinition,
   ]);
 
+  const onCancel = useCallback(() => {
+    setIsModalOpen(false);
+  }, []);
+
   return (
-    <Button
-      icon={<ImportIcon />}
-      variant="plain"
-      title={hasSchema ? 'Update schema' : 'Attach a schema'}
-      aria-label={hasSchema ? 'Update schema' : 'Attach schema'}
-      data-testid={`attach-schema-${documentType}-${documentId}-button`}
-      onClick={onClick}
-    />
+    <>
+      <Button
+        icon={<ImportIcon />}
+        variant="plain"
+        title={`${actionName} schema`}
+        aria-label={`${actionName} schema`}
+        data-testid={`attach-schema-${documentType}-${documentId}-button`}
+        onClick={onModalOpen}
+      />
+      <Modal isOpen={isModalOpen} variant="small" data-testid="attach-schema-modal">
+        <ModalHeader title={`${actionName} schema : ( ${documentTypeLabel} )`} />
+        <ModalBody>
+          <Stack hasGutter>
+            <StackItem>
+              <InputGroup>
+                <InputGroupItem isFill>
+                  <TextInput
+                    type="text"
+                    aria-label="Attaching schema file name"
+                    data-testid="attach-schema-modal-text"
+                    readOnly
+                    value={filePaths.join(', ')}
+                  />
+                </InputGroupItem>
+                <InputGroupItem>
+                  <Button data-testid="attach-schema-modal-btn-file" icon={<FileImportIcon />} onClick={onFileUpload} />
+                </InputGroupItem>
+              </InputGroup>
+            </StackItem>
+            <StackItem>
+              <InputGroup>
+                <InputGroupItem>
+                  <Radio
+                    id="option-xml-schema"
+                    name="schema-type"
+                    aria-label="XML Schema"
+                    label="XML Schema"
+                    data-testid="attach-schema-modal-option-xml"
+                    isChecked={selectedSchemaType === DocumentDefinitionType.XML_SCHEMA}
+                    onChange={() => setSelectedSchemaType(DocumentDefinitionType.XML_SCHEMA)}
+                  />
+                </InputGroupItem>
+                {documentType !== DocumentType.SOURCE_BODY && (
```

Comment:
Should we see JSON here? So far I only see XML, although I can pick a JSON schema and it works.

<img width="569" height="243" alt="Image" src="https://github.com/user-attachments/assets/18f98abe-6ec6-47ce-852e-0e1318908359" />

Reply:
Ah, it seems it's only for the target, is that intended?

<img width="569" height="243" alt="Image" src="https://github.com/user-attachments/assets/7a3b142b-6169-4d8c-bdda-7e0f8199ed91" />

Reply:
ugh thanks this is a good catch, it shouldn't allow importing JSON schema for the source body...
Although the file picker is filtered with extension, user can still choose "All Files" and upload JSON schema. This is a bug.

Aside from that, due to the limitation on `camel-xslt-saxon`, the source body has to be XML. In the future, we might either
1. Enhance `camel-xslt-saxon` to accept JSON body with custom input reader (to be confirmed it's feasible...)
2. Enhance Kaoto DataMapper step group to pre-process JSON body, for example, move it into an exchange variable before the XSLT step automatically

Reply:
https://github.com/KaotoIO/kaoto/issues/2429